components/server/src/code-sync/code-sync-service.ts

```diff
@@ -0,0 +1,214 @@
+        router.use((_, res, next) => {
+            // to correlate errors reported by users with errors logged by the server
+            res.setHeader('x-operation-id', uuid.v4());
+            return next();
+        });
+        router.use(this.auth.restHandler);
+        router.use((req, res, next) => {
+            if (!isWithFunctionAccessGuard(req) || !req.functionGuard?.canAccess('accessCodeSyncStorage')) {
+                res.sendStatus(403);
+                return;
+            }
+            return next();
+        });
+        router.use(bodyParser.text());
+        router.get('/v1/manifest', async (req, res) => {
+            if (!req.isAuthenticated() || !User.is(req.user)) {
+                res.sendStatus(401);
+                return;
+            }
+            const manifest = await this.db.getManifest(req.user.id);
+            res.json(manifest);
+            return;
+        });
+        router.get('/v1/resource/:resource', async (req, res) => {
+            if (!req.isAuthenticated() || !User.is(req.user)) {
+                res.sendStatus(401);
+                return;
+            }
+            const resourceKey = ALL_SERVER_RESOURCES.find(key => key === req.params.resource);
+            const revs = resourceKey && await this.db.getResources(req.user.id, resourceKey)
+            if (!revs || !revs.length) {
+                res.sendStatus(204);
+                return;
+            }
+            const result: { url: string, created: number }[] = revs.map(e => ({
+                url: req.originalUrl + '/' + e.rev,
+                created: Date.parse(e.created) / 1000 /* client expects in secondsm */
+            }));
+            res.json(result);
+            return;
+        });
+        router.get('/v1/resource/:resource/:ref', async (req, res) => {
+            if (!req.isAuthenticated() || !User.is(req.user)) {
+                res.sendStatus(401);
+                return;
+            }
+            const resourceKey = ALL_SERVER_RESOURCES.find(key => key === req.params.resource);
+            if (!resourceKey) {
+                res.sendStatus(204);
+                return;
+            }
+            const resource = await this.db.getResource(req.user.id, resourceKey, req.params.ref);
+            if (!resource) {
+                res.setHeader('etag', '0');
+                res.sendStatus(204);
+                return;
+            }
+            if (req.headers['If-None-Match'] === resource.rev) {
+                res.sendStatus(304);
+                return;
+            }
+
+            const request = new DownloadUrlRequest();
+            request.setOwnerId(req.user.id);
+            request.setName(toObjectName(resourceKey, resource.rev));
+            try {
+                const urlResponse = await util.promisify<DownloadUrlRequest, DownloadUrlResponse>(this.blobs.downloadUrl.bind(this.blobs))(request);
+                const response = await fetch(urlResponse.getUrl());
```

Comment:
not sure about ETag and Content Type, but the URLs returned by content service can be passed out, i.e. you could redirect to that URL rather than pass it through server (less bandwidth/load on the server).

Reply:
I will check whether Code can handle redirects.

Reply:
Unfortunately Code does not follow redirects here, we will have to do it in the ngix proxy with Lua scripts. I will skip it for now.